In [37]:
import pandas as pd

SyntaxError: invalid syntax (1139218586.py, line 2)

In [38]:
pip install pyjanitor

Defaulting to user installation because normal site-packages is not writeable
     -------------------------------------- 158.4/158.4 kB 1.9 MB/s eta 0:00:00
Note: you may need to restart the kernel to use updated packages.


  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.


In [39]:
from janitor import clean_names

# MVP

## Joins

### Question 1

Read in all 4 credit card transaction datasets and clean up the column names.

- CardBase.csv
- CustomerBase.csv
- FraudBase.csv
- TransactionBase.csv

In [2]:
card_base = pd.read_csv("data/CardBase.csv")
customer_base = pd.read_csv("data/CustomerBase.csv")
fraud_base = pd.read_csv("data/FraudBase.csv")
transaction_base = pd.read_csv("data/TransactionBase.csv")

### Question 2

Join the data containing card details and customer details by customer id, so that all records of card details and any matching records in customer details are kept.

In [3]:
card_base

Card_Number Card_Family  Credit_Limit  Cust_ID
0    8638-5407-3631-8196     Premium        530000  CC67088
1    7106-4239-7093-1515        Gold         18000  CC12076
2    6492-5655-8241-3530     Premium        596000  CC97173
3    2868-5606-5152-5706        Gold         27000  CC55858
4    1438-6906-2509-8219    Platinum        142000  CC90518
..                   ...         ...           ...      ...
495  4708-4407-9601-6022     Premium        691000  CC64993
496  6042-2856-7280-2925        Gold         33000  CC26787
497  8706-3809-5167-3899     Premium        144000  CC32532
498  9540-8558-5897-5046     Premium        830000  CC90246
499  5587-7265-8118-7718        Gold         24000  CC37803

[500 rows x 4 columns]

In [4]:
customer_base

Cust_ID  Age Customer_Segment Customer_Vintage_Group
0     CC25034   35          Diamond                    VG1
1     CC59625   22          Diamond                    VG1
2     CC69314   33          Diamond                    VG1
3     CC67036   46          Diamond                    VG1
4     CC25597   44          Diamond                    VG1
...       ...  ...              ...                    ...
5669  CC53430   27         Platinum                    VG2
5670  CC31823   33         Platinum                    VG2
5671  CC15335   28         Platinum                    VG2
5672  CC81556   26         Platinum                    VG2
5673  CC38634   21         Platinum                    VG2

[5674 rows x 4 columns]

In [5]:
card_base.merge(customer_base, how = "left", on = "Cust_ID")

Card_Number Card_Family  Credit_Limit  Cust_ID  Age  \
0    8638-5407-3631-8196     Premium        530000  CC67088   27   
1    7106-4239-7093-1515        Gold         18000  CC12076   48   
2    6492-5655-8241-3530     Premium        596000  CC97173   22   
3    2868-5606-5152-5706        Gold         27000  CC55858   30   
4    1438-6906-2509-8219    Platinum        142000  CC90518   50   
..                   ...         ...           ...      ...  ...   
495  4708-4407-9601-6022     Premium        691000  CC64993   34   
496  6042-2856-7280-2925        Gold         33000  CC26787   46   
497  8706-3809-5167-3899     Premium        144000  CC32532   28   
498  9540-8558-5897-5046     Premium        830000  CC90246   35   
499  5587-7265-8118-7718        Gold         24000  CC37803   21   

    Customer_Segment Customer_Vintage_Group  
0            Diamond                    VG1  
1               Gold                    VG3  
2            Diamond                    VG1  
3            Diamond                    VG1  
4           Platinum                    VG2  
..               ...                    ...  
495          Diamond                    VG1  
496         Platinum                    VG2  
497         Platinum                    VG2  
498             Gold                    VG3  
499          Diamond                    VG1  

[500 rows x 7 columns]

### Question 3

Join the data containing fraud details with transaction details so all rows of both tables are kept. What does the resulting row number tell you?

In [6]:
fraud_base

Transaction_ID  Fraud_Flag
0     CTID50558449           1
1     CTID55936882           1
2     CTID63762180           1
3     CTID76723439           1
4     CTID21246201           1
..             ...         ...
104   CTID57993591           1
105   CTID91108283           1
106   CTID30494187           1
107   CTID51301522           1
108   CTID49517337           1

[109 rows x 2 columns]

In [7]:
transaction_base

Transaction_ID Transaction_Date       Credit_Card_ID  Transaction_Value  \
0      CTID28830551        24-Apr-16  1629-9566-3285-2123              23649   
1      CTID45504917        11-Feb-16  3697-6001-4909-5350              26726   
2      CTID47312290         1-Nov-16  5864-4475-3659-1440              22012   
3      CTID25637718        28-Jan-16  5991-4421-8476-3804              37637   
4      CTID66743960        17-Mar-16  1893-8853-9900-8478               5113   
...             ...              ...                  ...                ...   
9995   CTID25037573        10-Jun-16  9157-2802-8374-1145              44280   
9996   CTID43832298        23-May-16  7416-4529-6690-5703              27163   
9997   CTID56532072         6-Aug-16  6699-2639-4522-6219              36424   
9998   CTID88101446        10-Oct-16  8341-5263-4582-7396               8464   
9999   CTID64730271        18-Jun-16  1947-8602-1695-7503               7606   

     Transaction_Segment  
0                  SEG25  
1                  SEG16  
2                  SEG14  
3                  SEG17  
4                  SEG14  
...                  ...  
9995               SEG18  
9996               SEG12  
9997               SEG13  
9998               SEG14  
9999               SEG15  

[10000 rows x 5 columns]

In [8]:
fraud_base.merge(transaction_base, how = "outer", on = "Transaction_ID")

#the resulting row is 10,000 which is the same as our transactions_base. 
#This means that all fraud cases have a matching transaction

Transaction_ID  Fraud_Flag Transaction_Date       Credit_Card_ID  \
0      CTID50558449         1.0         6-May-16  4246-1369-3659-8804   
1      CTID55936882         1.0        29-Nov-16  1336-9200-1264-2551   
2      CTID63762180         1.0         5-Dec-16  8528-6154-7390-5081   
3      CTID76723439         1.0        15-Sep-16  7908-2695-7391-7499   
4      CTID21246201         1.0        29-Feb-16  2524-4184-5908-6750   
...             ...         ...              ...                  ...   
9995   CTID25037573         NaN        10-Jun-16  9157-2802-8374-1145   
9996   CTID43832298         NaN        23-May-16  7416-4529-6690-5703   
9997   CTID56532072         NaN         6-Aug-16  6699-2639-4522-6219   
9998   CTID88101446         NaN        10-Oct-16  8341-5263-4582-7396   
9999   CTID64730271         NaN        18-Jun-16  1947-8602-1695-7503   

      Transaction_Value Transaction_Segment  
0                  6984               SEG23  
1                 34367               SEG21  
2                 44550               SEG15  
3                 48275               SEG16  
4                 35751               SEG25  
...                 ...                 ...  
9995              44280               SEG18  
9996              27163               SEG12  
9997              36424               SEG13  
9998               8464               SEG14  
9999               7606               SEG15  

[10000 rows x 6 columns]

### Question 4

**[Harder]** Join the data containing card details with transaction details so rows from the first which have matching ones in 
the second are returned.

Hint: `semi` is not an allowed option for `how` in `.merge()`. You'll need to think about the process of a semi 
join and then perform each step.

In [9]:
transaction_base

Transaction_ID Transaction_Date       Credit_Card_ID  Transaction_Value  \
0      CTID28830551        24-Apr-16  1629-9566-3285-2123              23649   
1      CTID45504917        11-Feb-16  3697-6001-4909-5350              26726   
2      CTID47312290         1-Nov-16  5864-4475-3659-1440              22012   
3      CTID25637718        28-Jan-16  5991-4421-8476-3804              37637   
4      CTID66743960        17-Mar-16  1893-8853-9900-8478               5113   
...             ...              ...                  ...                ...   
9995   CTID25037573        10-Jun-16  9157-2802-8374-1145              44280   
9996   CTID43832298        23-May-16  7416-4529-6690-5703              27163   
9997   CTID56532072         6-Aug-16  6699-2639-4522-6219              36424   
9998   CTID88101446        10-Oct-16  8341-5263-4582-7396               8464   
9999   CTID64730271        18-Jun-16  1947-8602-1695-7503               7606   

     Transaction_Segment  
0                  SEG25  
1                  SEG16  
2                  SEG14  
3                  SEG17  
4                  SEG14  
...                  ...  
9995               SEG18  
9996               SEG12  
9997               SEG13  
9998               SEG14  
9999               SEG15  

[10000 rows x 5 columns]

In [10]:
(
card_base
    .merge(transaction_base[["Credit_Card_ID"]], how = "inner", left_on = "Card_Number", right_on = "Credit_Card_ID")
    .drop_duplicates()
)

Card_Number Card_Family  Credit_Limit  Cust_ID  \
0     8638-5407-3631-8196     Premium        530000  CC67088   
23    7106-4239-7093-1515        Gold         18000  CC12076   
32    6492-5655-8241-3530     Premium        596000  CC97173   
53    2868-5606-5152-5706        Gold         27000  CC55858   
62    1438-6906-2509-8219    Platinum        142000  CC90518   
...                   ...         ...           ...      ...   
9893  4708-4407-9601-6022     Premium        691000  CC64993   
9914  6042-2856-7280-2925        Gold         33000  CC26787   
9937  8706-3809-5167-3899     Premium        144000  CC32532   
9961  9540-8558-5897-5046     Premium        830000  CC90246   
9978  5587-7265-8118-7718        Gold         24000  CC37803   

           Credit_Card_ID  
0     8638-5407-3631-8196  
23    7106-4239-7093-1515  
32    6492-5655-8241-3530  
53    2868-5606-5152-5706  
62    1438-6906-2509-8219  
...                   ...  
9893  4708-4407-9601-6022  
9914  6042-2856-7280-2925  
9937  8706-3809-5167-3899  
9961  9540-8558-5897-5046  
9978  5587-7265-8118-7718  

[500 rows x 5 columns]

## Tidy Data

### Question 5

Read in hat_observations and split `observation` into two columns, `hat_colour` and `hat_type`.

Hint: Look up the pandas documentation for `.str.split()`

In [41]:
hat_observations = pd.read_csv("data/hat_observations.csv")
hat_observations

day  month  year        observation  observation_count
0    25     10  2019        black,beret                  7
1    11      9  2019       blue,top hat                 10
2     4     12  2018    green,santa hat                  9
3    18      6  2018        white,beret                  9
4    26      9  2018    black,santa hat                  1
..  ...    ...   ...                ...                ...
95   15      8  2018  purple,bucket hat                  8
96   26      1  2018   green,bowler hat                  5
97   15     11  2018      black,top hat                  7
98   14      9  2019       green,helmet                  4
99    1     10  2018        blue,fedora                  4

[100 rows x 5 columns]

In [42]:
hat_observations[['hat_colour', 'hat_type']] = hat_observations.observation.str.split(',', expand=True)
hat_observations.drop(columns = "observation", inplace = True)
hat_observations

day  month  year  observation_count hat_colour    hat_type
0    25     10  2019                  7      black       beret
1    11      9  2019                 10       blue     top hat
2     4     12  2018                  9      green   santa hat
3    18      6  2018                  9      white       beret
4    26      9  2018                  1      black   santa hat
..  ...    ...   ...                ...        ...         ...
95   15      8  2018                  8     purple  bucket hat
96   26      1  2018                  5      green  bowler hat
97   15     11  2018                  7      black     top hat
98   14      9  2019                  4      green      helmet
99    1     10  2018                  4       blue      fedora

[100 rows x 6 columns]

### Question 6

Unite day, month, and year columns into a column called date using a suitable separator. Then find the date where
the most berets were observed.

Hints:
- you can use the `.agg()` function to combine columns
- you may need to convert the data types of the columns you wish to combine

In [13]:
hat_observations.dtypes

day                   int64
month                 int64
year                  int64
observation_count     int64
hat_colour           object
hat_type             object
dtype: object

In [43]:
hat_observations["year"] = hat_observations["year"].astype(str)

In [44]:
hat_observations["month"] = hat_observations["month"].astype(str)

In [45]:
hat_observations["day"] = hat_observations["day"].astype(str)

In [46]:
hat_observations

day month  year  observation_count hat_colour    hat_type
0   25    10  2019                  7      black       beret
1   11     9  2019                 10       blue     top hat
2    4    12  2018                  9      green   santa hat
3   18     6  2018                  9      white       beret
4   26     9  2018                  1      black   santa hat
..  ..   ...   ...                ...        ...         ...
95  15     8  2018                  8     purple  bucket hat
96  26     1  2018                  5      green  bowler hat
97  15    11  2018                  7      black     top hat
98  14     9  2019                  4      green      helmet
99   1    10  2018                  4       blue      fedora

[100 rows x 6 columns]

In [47]:
hat_observations['date'] = hat_observations[['day', 'month', 'year']].agg('-'.join, axis = 1)

hat_observations

day month  year  observation_count hat_colour    hat_type        date
0   25    10  2019                  7      black       beret  25-10-2019
1   11     9  2019                 10       blue     top hat   11-9-2019
2    4    12  2018                  9      green   santa hat   4-12-2018
3   18     6  2018                  9      white       beret   18-6-2018
4   26     9  2018                  1      black   santa hat   26-9-2018
..  ..   ...   ...                ...        ...         ...         ...
95  15     8  2018                  8     purple  bucket hat   15-8-2018
96  26     1  2018                  5      green  bowler hat   26-1-2018
97  15    11  2018                  7      black     top hat  15-11-2018
98  14     9  2019                  4      green      helmet   14-9-2019
99   1    10  2018                  4       blue      fedora   1-10-2018

[100 rows x 7 columns]

In [48]:
hat_observations = hat_observations.loc[:, ["observation_count", "hat_colour", "hat_type", "date"]]
hat_observations

observation_count hat_colour    hat_type        date
0                   7      black       beret  25-10-2019
1                  10       blue     top hat   11-9-2019
2                   9      green   santa hat   4-12-2018
3                   9      white       beret   18-6-2018
4                   1      black   santa hat   26-9-2018
..                ...        ...         ...         ...
95                  8     purple  bucket hat   15-8-2018
96                  5      green  bowler hat   26-1-2018
97                  7      black     top hat  15-11-2018
98                  4      green      helmet   14-9-2019
99                  4       blue      fedora   1-10-2018

[100 rows x 4 columns]

In [40]:
%who

card_base	 clean_names	 customer_base	 fraud_base	 hat_observations	 pd	 transaction_base	 


In [49]:
(
    hat_observations
    .loc[(hat_observations['hat_type'] == 'beret')]
    .groupby('date', as_index=False)
    .agg(observations = ('observation_count', 'sum'))
    .sort_values('observations')
    .tail(1)
)

date  observations
2  18-6-2018            11

# Extension

## Joins

### Question 1
Can you join all 4 credit card datasets together so that you’re left with a dataset that looks like below with 109 rows and 12 columns?

![](images/all_joined.png)

## Tidy Data

### Question 2

Read in exam_scores and transform it into long format with two new columns `exam_question` and `score`. Remove 
superfluous information from the values in `exam_question`.